subprojects/gst-plugins-good/gst/rtp/gstrtph266pay.c

```diff
@@ -590,48 +659,142 @@ _set_au_boundaries (GstRtpH266Pay * rtph266pay)
+      if (ret != GST_FLOW_OK)
+        return ret;
+    } else {                    // Not aggregation path
+      gboolean is_last = g_queue_get_length (nalus) == 0;
+      // Can't push the last NALU without knowing if it's the end of an AU,
+      // because setting the M bit could be necessary. But have to push it if
+      // we're on EOS
+      if (!eos && is_last && !nalu->au_end) {
+        GST_DEBUG_OBJECT (rtph266pay, "Keeping last NALU: %" NALU_PTR_FORMAT,
+            NALU_ARGS (nalu));
+        g_queue_push_head (nalus, nalu);
+        return GST_FLOW_OK;
+      }
+
+      // Push single Unit Packet or several Fragmented Units
+      ret = _push_single (rtph266pay, nalu);
+      if (ret != GST_FLOW_OK)
+        return ret;
+    }
+  }
+
+  if (_aggregated_nalus_have_au (aggregated_nalus))
+    return _push_aggregated (rtph266pay);
+  return GST_FLOW_OK;
+}
+
+static GstFlowReturn
+_push_single (GstRtpH266Pay * rtph266pay, Nalu * nalu)
+{
+  GstFlowReturn ret;
+
+  GST_DEBUG_OBJECT (rtph266pay, "Pushing NALU: %" NALU_PTR_FORMAT,
+      NALU_ARGS (nalu));
+
+  if (_up_fits_in_mtu (rtph266pay, nalu))
+    ret = _push_unit_pkt (rtph266pay, nalu);
+  else
+    ret = _push_fragmented (rtph266pay, nalu);
+
+  return ret;
+}
+
+static GstFlowReturn
+_push_aggregated (GstRtpH266Pay * rtph266pay)
+{
+  GstRTPBasePayload *rtpbasepay = GST_RTP_BASE_PAYLOAD (rtph266pay);
+  AggregatedNalus *aggregated_nalus = &rtph266pay->aggregated_nalus;
+  GQueue *agg_nalus = &aggregated_nalus->nalus;
+  GstRTPBuffer rtp = GST_RTP_BUFFER_INIT;
+  GstBuffer *out_buf;
+  guint num_nalus;
+  Nalu *nalu;
 
-    if (_up_fits_in_mtu (rtph266pay, nalu))
-      ret = _push_up (rtph266pay, nalu);
-    else
-      ret = _push_fu (rtph266pay, nalu);
+  // Can't send an AP with a single NALU
+  num_nalus = g_queue_get_length (agg_nalus);
+  if (num_nalus == 1) {
+    GST_DEBUG_OBJECT (rtph266pay, "Can't send an AP with a single NALU");
+    nalu = g_queue_pop_head (agg_nalus);
+    _aggregated_nalus_clear (aggregated_nalus);
+    return _push_unit_pkt (rtph266pay, nalu);
+  }
 
+  // Allocate just the RTP header. We'll add the buffer payload later. This way
+  // we avoid unnecessary copies
+  out_buf = gst_rtp_base_payload_allocate_output_buffer (rtpbasepay, 0, 0, 0);
+  if (!gst_rtp_buffer_map (out_buf, GST_MAP_WRITE, &rtp))
+    goto error;
+
+  // Copy buffer metadata from the last nalu to aggregate
+  nalu = g_queue_peek_tail (agg_nalus);
+  gst_buffer_copy_into (out_buf, nalu->nalu_buf,
+      GST_BUFFER_COPY_FLAGS | GST_BUFFER_COPY_TIMESTAMPS, 0, -1);
+  gst_rtp_buffer_set_marker (&rtp, nalu->au_end);
+
+  // Append PayloadHdr
+  // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+  // |F|Z| MaxLayerID| Type(28)| TID |
+  // +---------------+---------------+
+  guint16 f_bit_msk = ((aggregated_nalus->f_bit) << (7 + 8)) & 0x70;
+  guint16 max_layer_id_msk = ((aggregated_nalus->max_layer_id) << 8) & 0x3F;
```

Comment:
Shouldn't be `((aggregated_nalus->max_layer_id & 0x3F) << 8);`?

Reply:
You're right. Good catch!